File: android/app/src/main/java/com/droppix/app/protocol/Protocol.kt
package com.droppix.app.protocol

enum class MsgType(val code: Int) {
    HELLO(1), CONFIG(2), VIDEO(3), PING(4), PONG(5), BYE(6), INPUT(7), ORIENTATION(8),
    AUDIO(9), OVERLAY(10), TOUCH(11), SCROLL(12), MOUSE_BUTTON(13), KEY(14);
    companion object {
        fun fromCode(c: Int): MsgType? = entries.firstOrNull { it.code == c }
    }
}

data class ParsedMessage(val type: MsgType, val body: ByteArray)

// One finger in a multi-touch report: id (stable across a gesture), x/y in 0..65535 across
// the display, pressure in 0..1023.
data class Contact(val id: Int, val x: Int, val y: Int, val pressure: Int)

object Protocol {
    const val VERSION = 5

    private fun putU32(out: ArrayList<Byte>, x: Int) {
        out.add((x ushr 24).toByte()); out.add((x ushr 16).toByte())
        out.add((x ushr 8).toByte()); out.add(x.toByte())
    }
    private fun getU32(b: ByteArray, o: Int): Int =
        ((b[o].toInt() and 0xFF) shl 24) or ((b[o+1].toInt() and 0xFF) shl 16) or
        ((b[o+2].toInt() and 0xFF) shl 8) or (b[o+3].toInt() and 0xFF)
    private fun getU64(b: ByteArray, o: Int): Long {
        var x = 0L
        for (i in 0 until 8) x = (x shl 8) or (b[o + i].toLong() and 0xFF)
        return x
    }

    fun encodeMessage(type: MsgType, body: ByteArray): ByteArray {
        val out = ArrayList<Byte>(5 + body.size)
        putU32(out, 1 + body.size)            // length covers type + body
        out.add(type.code.toByte())
        for (b in body) out.add(b)
        return out.toByteArray()
    }

    fun encodeHello(version: Int, width: Int, height: Int, density: Int,
                    name: String = "", id: String = "",
                    fps: Int = 0, audioWanted: Int = 0, orientationCode: Int = 0, bitrateKbps: Int = 0): ByteArray {
        val out = ArrayList<Byte>()
        putU32(out, version); putU32(out, width); putU32(out, height); putU32(out, density)
        if (version >= 4) {
            putU32(out, fps); out.add(audioWanted.toByte()); out.add(orientationCode.toByte())
        }
        if (version >= 5) { putU32(out, bitrateKbps) }
        val n = name.toByteArray(Charsets.UTF_8); val i = id.toByteArray(Charsets.UTF_8)
        out.add((n.size ushr 8).toByte()); out.add(n.size.toByte()); for (x in n) out.add(x)
        out.add((i.size ushr 8).toByte()); out.add(i.size.toByte()); for (x in i) out.add(x)
        return out.toByteArray()
    }

    // INPUT body: u8 action (0=down,1=move,2=up), u16 x_norm, u16 y_norm, u16 pressure (big-endian).
    fun encodeInput(action: Int, xNorm: Int, yNorm: Int, pressure: Int): ByteArray {
        val out = ArrayList<Byte>(7)
        out.add(action.toByte())
        out.add((xNorm ushr 8).toByte()); out.add(xNorm.toByte())
        out.add((yNorm ushr 8).toByte()); out.add(yNorm.toByte())
        out.add((pressure ushr 8).toByte()); out.add(pressure.toByte())
        return out.toByteArray()
    }

    // TOUCH body: u8 count, then count x { u8 id, u16 x, u16 y, u16 pressure } (big-endian).
    // The full set of active contacts each event (count 0 = all up); capped at 10.
    fun encodeTouch(contacts: List<Contact>): ByteArray {
        val n = minOf(contacts.size, 10)
        val out = ArrayList<Byte>(1 + n * 7)
        out.add(n.toByte())
        for (i in 0 until n) {
            val c = contacts[i]
            out.add(c.id.toByte())
            out.add((c.x ushr 8).toByte()); out.add(c.x.toByte())
            out.add((c.y ushr 8).toByte()); out.add(c.y.toByte())
            out.add((c.pressure ushr 8).toByte()); out.add(c.pressure.toByte())
        }
        return out.toByteArray()
    }

    // ORIENTATION body: u8 code (0=0°, 1=90°, 2=180°, 3=270°).
    fun encodeOrientation(code: Int): ByteArray = byteArrayOf(code.toByte())

    // SCROLL body: i16 dx, i16 dy, u16 x, u16 y (big-endian).
    fun encodeScroll(dx: Int, dy: Int, x: Int, y: Int): ByteArray {
        val out = ArrayList<Byte>(8)
        fun u16(v: Int) { out.add((v ushr 8).toByte()); out.add(v.toByte()) }
        u16(dx); u16(dy); u16(x); u16(y)
        return out.toByteArray()
    }

    // MOUSE_BUTTON body: u8 button, u8 action, u16 x, u16 y (big-endian).
    fun encodeMouseButton(button: Int, action: Int, x: Int, y: Int): ByteArray {
        val out = ArrayList<Byte>(6)
        out.add(button.toByte()); out.add(action.toByte())
        out.add((x ushr 8).toByte()); out.add(x.toByte())
        out.add((y ushr 8).toByte()); out.add(y.toByte())
        return out.toByteArray()
    }

    // KEY body: u16 keycode (big-endian), u8 action (0=up,1=down,2=repeat).
    fun encodeKey(keycode: Int, action: Int): ByteArray =
        byteArrayOf((keycode ushr 8).toByte(), keycode.toByte(), action.toByte())

    data class Config(val width: Int, val height: Int, val fps: Int, val extradata: ByteArray)
    fun decodeConfig(body: ByteArray): Config? {
        if (body.size < 16) return null
        val w = getU32(body, 0); val h = getU32(body, 4); val fps = getU32(body, 8)
        val n = getU32(body, 12)
        if (body.size != 16 + n) return null
        return Config(w, h, fps, body.copyOfRange(16, body.size))
    }

    data class Video(val ptsUs: Long, val keyframe: Boolean, val nal: ByteArray)
    fun decodeVideo(body: ByteArray): Video? {
        if (body.size < 9) return null
        val pts = getU64(body, 0)
        val key = body[8].toInt() != 0
        return Video(pts, key, body.copyOfRange(9, body.size))
    }
}

class MessageParser {
    private var buf = ByteArray(0)
    private var pos = 0
    private val maxMessage = 64 * 1024 * 1024  // mirror host sanity cap

    fun feed(data: ByteArray, n: Int) {
        buf = buf.plus(data.copyOfRange(0, n))  // buf.plus already allocates a fresh array
    }

    fun next(): ParsedMessage? {
        while (true) {
            if (buf.size - pos < 4) return null
            val len = ((buf[pos].toInt() and 0xFF) shl 24) or
                      ((buf[pos+1].toInt() and 0xFF) shl 16) or
                      ((buf[pos+2].toInt() and 0xFF) shl 8) or
                      (buf[pos+3].toInt() and 0xFF)
            if (len < 1 || len > maxMessage) { pos += 4; continue }
            if (buf.size - pos < 4 + len) return null
            val type = MsgType.fromCode(buf[pos + 4].toInt() and 0xFF)
            val body = buf.copyOfRange(pos + 5, pos + 4 + len)
            pos += 4 + len
            if (pos > 65536) { buf = buf.copyOfRange(pos, buf.size); pos = 0 }
            if (type == null) continue   // unknown type: skip
            return ParsedMessage(type, body)
        }
    }
}

File: host/src/protocol.cpp
#include "protocol.h"

namespace droppix {
namespace {

void put_u32(std::vector<unsigned char>& v, uint32_t x) {
  v.push_back((x >> 24) & 0xFF); v.push_back((x >> 16) & 0xFF);
  v.push_back((x >> 8) & 0xFF);  v.push_back(x & 0xFF);
}
void put_u64(std::vector<unsigned char>& v, uint64_t x) {
  for (int s = 56; s >= 0; s -= 8) v.push_back((x >> s) & 0xFF);
}
uint32_t get_u32(const unsigned char* p) {
  return (uint32_t(p[0]) << 24) | (uint32_t(p[1]) << 16) |
         (uint32_t(p[2]) << 8) | uint32_t(p[3]);
}
uint64_t get_u64(const unsigned char* p) {
  uint64_t x = 0;
  for (int i = 0; i < 8; ++i) x = (x << 8) | p[i];
  return x;
}
void put_u16(std::vector<unsigned char>& v, uint16_t x) {
  v.push_back((x >> 8) & 0xFF); v.push_back(x & 0xFF);
}
uint16_t get_u16(const unsigned char* p) {
  return (uint16_t(p[0]) << 8) | uint16_t(p[1]);
}

}  // namespace

std::vector<unsigned char> encode_message(MsgType type,
                                          const std::vector<unsigned char>& body) {
  std::vector<unsigned char> m;
  uint32_t len = 1 + static_cast<uint32_t>(body.size());
  put_u32(m, len);
  m.push_back(static_cast<unsigned char>(type));
  m.insert(m.end(), body.begin(), body.end());
  return m;
}

void MessageParser::feed(const unsigned char* data, size_t n) {
  buf_.insert(buf_.end(), data, data + n);
}

bool MessageParser::next(ParsedMessage& out) {
  static constexpr uint32_t kMaxMessage = 64u * 1024u * 1024u;  // 64 MiB sanity cap
  for (;;) {
    if (buf_.size() - pos_ < 4) return false;
    uint32_t len = get_u32(buf_.data() + pos_);
    if (len < 1 || len > kMaxMessage) {
      pos_ += 4;       // malformed length word: skip and resync (no recursion)
      continue;
    }
    if (buf_.size() - pos_ < static_cast<size_t>(4) + len) return false;  // widened
    const unsigned char* p = buf_.data() + pos_ + 4;
    out.type = static_cast<MsgType>(p[0]);
    out.body.assign(p + 1, p + len);
    pos_ += static_cast<size_t>(4) + len;
    if (pos_ > 65536) { buf_.erase(buf_.begin(), buf_.begin() + pos_); pos_ = 0; }
    return true;
  }
}

std::vector<unsigned char> encode_hello(uint32_t version, uint32_t w, uint32_t h,
                                        uint32_t d, const std::string& name, const std::string& id,
                                        uint32_t fps, uint8_t audio_wanted, uint8_t orientation_code,
                                        uint32_t bitrate_kbps) {
  std::vector<unsigned char> b;
  put_u32(b, version); put_u32(b, w); put_u32(b, h); put_u32(b, d);
  if (version >= 4) {
    put_u32(b, fps); b.push_back(audio_wanted); b.push_back(orientation_code);
  }
  if (version >= 5) {
    put_u32(b, bitrate_kbps);
  }
  put_u16(b, (uint16_t)name.size()); b.insert(b.end(), name.begin(), name.end());
  put_u16(b, (uint16_t)id.size());   b.insert(b.end(), id.begin(),   id.end());
  return b;
}
bool decode_hello(const std::vector<unsigned char>& b, uint32_t& version,
                  uint32_t& w, uint32_t& h, uint32_t& d, uint32_t& fps,
                  uint8_t& audio_wanted, uint8_t& orientation_code,
                  uint32_t& bitrate_kbps, std::string& name, std::string& id) {
  if (b.size() < 16) return false;
  version = get_u32(b.data()); w = get_u32(b.data()+4);
  h = get_u32(b.data()+8); d = get_u32(b.data()+12);
  fps = 0; audio_wanted = 0; orientation_code = 0; bitrate_kbps = 0;
  name.clear(); id.clear();
  size_t p = 16;
  if (version >= 4) {
    if (b.size() < 22) return true;              // truncated v4 fixed block: keep sentinels
    fps = get_u32(b.data()+16); audio_wanted = b[20]; orientation_code = b[21];
    p = 22;
    if (version >= 5) {
      if (b.size() < 26) return true;            // truncated v5 fixed block: keep bitrate sentinel
      bitrate_kbps = get_u32(b.data()+22);
      p = 26;
    }
  }
  if (b.size() >= p+2) { uint16_t n = get_u16(b.data()+p); p += 2;
    if (b.size() >= p+n) { name.assign(b.begin()+p, b.begin()+p+n); p += n; } else return true; }
  if (b.size() >= p+2) { uint16_t n = get_u16(b.data()+p); p += 2;
    if (b.size() >= p+n) { id.assign(b.begin()+p, b.begin()+p+n); } }
  return true;
}
// Back-compat: thin forwarder for pre-v5 callers that don't need bitrate_kbps.
bool decode_hello(const std::vector<unsigned char>& b, uint32_t& version,
                  uint32_t& w, uint32_t& h, uint32_t& d, uint32_t& fps,
                  uint8_t& audio_wanted, uint8_t& orientation_code,
                  std::string& name, std::string& id) {
  uint32_t bitrate_kbps;
  return decode_hello(b, version, w, h, d, fps, audio_wanted, orientation_code,
                      bitrate_kbps, name, id);
}
bool decode_hello(const std::vector<unsigned char>& b, uint32_t& version,
                  uint32_t& w, uint32_t& h, uint32_t& d, std::string& name, std::string& id) {
  uint32_t fps; uint8_t audio, orient;
  return decode_hello(b, version, w, h, d, fps, audio, orient, name, id);
}

std::vector<unsigned char> encode_config(uint32_t w, uint32_t h, uint32_t fps,
                                         const std::vector<unsigned char>& ed) {
  std::vector<unsigned char> b;
  put_u32(b, w); put_u32(b, h); put_u32(b, fps);
  put_u32(b, static_cast<uint32_t>(ed.size()));
  b.insert(b.end(), ed.begin(), ed.end());
  return b;
}
bool decode_config(const std::vector<unsigned char>& b,
                   uint32_t& w, uint32_t& h, uint32_t& fps,
                   std::vector<unsigned char>& ed) {
  if (b.size() < 16) return false;
  w = get_u32(b.data()); h = get_u32(b.data() + 4); fps = get_u32(b.data() + 8);
  uint32_t n = get_u32(b.data() + 12);
  if (b.size() != static_cast<size_t>(16) + n) return false;
  ed.assign(b.begin() + 16, b.end());
  return true;
}

std::vector<unsigned char> encode_video(uint64_t pts_us, bool key,
                                        const std::vector<unsigned char>& nal) {
  std::vector<unsigned char> b;
  put_u64(b, pts_us);
  b.push_back(key ? 1 : 0);
  b.insert(b.end(), nal.begin(), nal.end());
  return b;
}
bool decode_video(const std::vector<unsigned char>& b,
                  uint64_t& pts_us, bool& key, std::vector<unsigned char>& nal) {
  if (b.size() < 9) return false;
  pts_us = get_u64(b.data());
  key = b[8] != 0;
  nal.assign(b.begin() + 9, b.end());
  return true;
}

std::vector<unsigned char> encode_input(uint8_t action, uint16_t x_norm, uint16_t y_norm,
                                        uint16_t pressure) {
  std::vector<unsigned char> b;
  b.push_back(action);
  put_u16(b, x_norm); put_u16(b, y_norm); put_u16(b, pressure);
  return b;
}
bool decode_input(const std::vector<unsigned char>& b,
                  uint8_t& action, uint16_t& x_norm, uint16_t& y_norm, uint16_t& pressure) {
  if (b.size() != 5 && b.size() != 7) return false;
  action = b[0];
  x_norm = get_u16(b.data() + 1);
  y_norm = get_u16(b.data() + 3);
  pressure = (b.size() == 7) ? get_u16(b.data() + 5) : 1023;  // 5-byte (old client) => full
  return true;
}

std::vector<unsigned char> encode_touch(const std::vector<TouchContact>& contacts) {
  std::vector<unsigned char> b;
  const size_t n = contacts.size() < 10 ? contacts.size() : 10;   // cap at 10 slots
  b.push_back(static_cast<unsigned char>(n));
  for (size_t i = 0; i < n; ++i) {
    b.push_back(contacts[i].id);
    put_u16(b, contacts[i].x);
    put_u16(b, contacts[i].y);
    put_u16(b, contacts[i].pressure);
  }
  return b;
}
bool decode_touch(const std::vector<unsigned char>& b, std::vector<TouchContact>& out) {
  out.clear();
  if (b.empty()) return false;
  const uint8_t n = b[0];
  if (b.size() != 1u + static_cast<size_t>(n) * 7u) return false;   // 7 bytes per contact
  const unsigned char* p = b.data() + 1;
  for (uint8_t i = 0; i < n; ++i, p += 7) {
    TouchContact c;
    c.id = p[0];
    c.x = get_u16(p + 1);
    c.y = get_u16(p + 3);
    c.pressure = get_u16(p + 5);
    out.push_back(c);
  }
  return true;
}

std::vector<unsigned char> encode_scroll(int16_t dx, int16_t dy, uint16_t x, uint16_t y) {
  std::vector<unsigned char> b;
  auto u16 = [&](uint16_t v){ b.push_back((unsigned char)(v >> 8)); b.push_back((unsigned char)(v & 0xFF)); };
  u16((uint16_t)dx); u16((uint16_t)dy); u16(x); u16(y);
  return b;
}
bool decode_scroll(const std::vector<unsigned char>& b, int16_t& dx, int16_t& dy,
                   uint16_t& x, uint16_t& y) {
  if (b.size() < 8) return false;
  auto u16 = [&](size_t o){ return (uint16_t)((b[o] << 8) | b[o+1]); };
  dx = (int16_t)u16(0); dy = (int16_t)u16(2); x = u16(4); y = u16(6);
  return true;
}

std::vector<unsigned char> encode_mouse_button(uint8_t button, uint8_t action,
                                               uint16_t x, uint16_t y) {
  return { button, action, (unsigned char)(x >> 8), (unsigned char)(x & 0xFF),
           (unsigned char)(y >> 8), (unsigned char)(y & 0xFF) };
}
bool decode_mouse_button(const std::vector<unsigned char>& b, uint8_t& button,
                         uint8_t& action, uint16_t& x, uint16_t& y) {
  if (b.size() < 6) return false;
  button = b[0]; action = b[1];
  x = (uint16_t)((b[2] << 8) | b[3]); y = (uint16_t)((b[4] << 8) | b[5]);
  return true;
}

std::vector<unsigned char> encode_key(uint16_t keycode, uint8_t action) {
  return { (unsigned char)(keycode >> 8), (unsigned char)(keycode & 0xFF), action };
}
bool decode_key(const std::vector<unsigned char>& b, uint16_t& keycode, uint8_t& action) {
  if (b.size() < 3) return false;
  keycode = (uint16_t)((b[0] << 8) | b[1]); action = b[2];
  return true;
}

std::vector<unsigned char> encode_pen(uint16_t x, uint16_t y, uint16_t pressure, uint8_t flags) {
  std::vector<unsigned char> b;
  put_u16(b, x); put_u16(b, y); put_u16(b, pressure); b.push_back(flags);
  return b;
}
bool decode_pen(const std::vector<unsigned char>& b, uint16_t& x, uint16_t& y,
                uint16_t& pressure, uint8_t& flags) {
  if (b.size() < 7) return false;
  x = (uint16_t)((b[0] << 8) | b[1]); y = (uint16_t)((b[2] << 8) | b[3]);
  pressure = (uint16_t)((b[4] << 8) | b[5]); flags = b[6];
  return true;
}

std::vector<unsigned char> encode_orientation(uint8_t code) {
  return {code};
}
bool decode_orientation(const std::vector<unsigned char>& b, uint8_t& code) {
  if (b.size() != 1) return false;
  code = b[0];
  return true;
}
std::vector<unsigned char> encode_overlay(uint8_t show) {
  return {show};
}
bool decode_overlay(const std::vector<unsigned char>& b, uint8_t& show) {
  if (b.size() != 1) return false;
  show = b[0];
  return true;
}

}  // namespace droppix
